libcudf Developer Guide [skip-ci] (#6977)
rapidsai/cudf: cpp/docs/DEVELOPER_GUIDE.md

host copies. Private member variables are typically prefixed with an underscore.

```c++
template <typename IteratorType>
void algorithm_function(int x, rmm::cuda_stream_view s, rmm::device_memory_resource* mr) 
{
  ...
}

class utility_class 
{
  ...
 private:
  int _rating{};
  std::unique_ptr<cudf::column> _column{};
}

TYPED_TEST_CASE(RepeatTypedTestFixture, cudf::test::FixedWidthTypes);

TYPED_TEST(RepeatTypedTestFixture, RepeatScalarCount)
{
  ...
}
```

C++ formatting is enforced using `clang-format`. You should configure `clang-format` on your 
machine to use the `cudf/cpp/.clang-format` configuration file, and run `clang-format` on all 
changed code before committing it. The easiest way to do this is to configure your editor to 
"format on save".

Aspects of code style not discussed in this document and not automatically enforceable are typically
caught during code review, or not enforced.

### C++ Guidelines

In general, we recommend following 
[C++ Core Guidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines). We also 
recommend watching Sean Parent's [C++ Seasoning talk](https://www.youtube.com/watch?v=W2tWOdzgXHA), 

> and we try to follow his rules: "No raw loops. No raw pointers. No raw synchronization primitives." 

Reviewer: AFAIK the pointer rule only refers to owning pointers. Passing pointers around without changing the ownership should be fine.
```suggestion
and we try to follow his rules: "No raw loops. No raw owning pointers. No raw synchronization primitives." 
```

Author: But it's a direct quote. And I think my bullet point after the quote clears this up. In general I don't think I agree. You will notice that we don't have raw pointers at all in our APIs unless necessary - this is what our view classes help with.

Reviewer: Did not realize it's literally a quote, disregard the comment then. Just trying to make the guideline clearer. 
Further down in the document we do recommend to pass in/out parameters as pointers (unless it's a type with a view).

 * Prefer algorithms from STL and Thrust to raw loops.

>  * Prefer libcudf and RMM [owning data structures and views](libcudf-data-tructures) to raw pointers

Reviewer 2: ```suggestion
 * Prefer libcudf and RMM [owning data structures and views](libcudf-data-structures) to raw pointers
```

kernels and device functions (including Thrust algorithms).

## `cudf::table`

Owning class for a set of `cudf::column`s all with equal number of elements. This is the C++ 
equivalent to a data frame. 

Implicitly convertible to `cudf::table_view` and `cudf::mutable_table_view`

Movable and copyable. A copy performs a deep copy of all columns, whereas a move moves all columns 
from one table to another.

### `cudf::table_view`

An *immutable*, non-owning view of a table. 

### `cudf::mutable_table_view`

A *mutable*, non-owning view of a table. 

## `cudf::scalar`

A `cudf::scalar` is an object that can represent a singular, nullable value of any of the types 
currently supported by cudf. Each type of value is represented by a separate type of scalar class 
which are all derived from `cudf::scalar`. e.g. A `numeric_scalar` holds a single numerical value, 
a `string_scalar` holds a single string. The data for the stored value resides in device memory.

|Value type|Scalar class|Notes|
|-|-|-|
|fixed-width|`fixed_width_scalar<T>`| `T` can be any fixed-width type|
|numeric|`numeric_scalar<T>` | `T` can be `int8_t`, `int16_t`, `int32_t`, `int_64_t`, `float` or `double`|

> |fixed-point|`fixed_point_scalar<T> | `T` can be `numeric::decimal32` or `numeric::decimal64`|

Reviewer 3: ```suggestion
|fixed-point|`fixed_point_scalar<T>` | `T` can be `numeric::decimal32` or `numeric::decimal64`|
```

Author: Great catch.

rmm::mr::device_memory_resource * mr = new my_custom_resource{...};
// Allocates uninitialized storage for 100 `int32_t` elements on stream `s` using the resource `mr`
rmm::device_uvector<int32_t> v2{100, s, mr}; 
```

## Input/Output Style<a name="inout_style"></a>

The preferred style for how inputs are passed in and outputs are returned is the following:
-   Inputs
	- Columns:
		- `column_view const&`
	- Tables:
		- `table_view const&`
    - Scalar:
        - `scalar const&`
    - Everything else:
       - Trivial or inexpensively copied types
          - Pass by value
       - Non-trivial or expensive to copy types
          - Pass by `const&`
-   In/Outs  
	- Columns:
		- `mutable_column_view&`
	- Tables:
		- `mutable_table_view&`
    - Everything else:
        - Pass by via raw pointer
-   Outputs 
	- Outputs should be *returned*, i.e., no output parameters
	- Columns:
		- `std::unique_ptr<column>`
	- Tables:
		- `std::unique_ptr<table>`
    - Scalars:
        - `std::unique_ptr<scalar>`

### Multiple Return Values

Sometimes it is necessary for functions to have multiple outputs. There are a few ways this can be 

> done in C++ (including creating a  `struct`  for the output). One convenient way to do this is 

Reviewer: Not sure if intentional
```suggestion
done in C++ (including creating a `struct` for the output). One convenient way to do this is 
```

using `std::tie`  and  `std::make_pair`. Note that objects passed to `std::make_pair` will invoke 
either the copy constructor or the move constructor of the object, and it may be preferable to move 
non-trivially copyable objects (and required for types with deleted copy constructors, like 
`std::unique_ptr`).

```c++
std::pair<table, table> return_two_tables(void){
  cudf::table out0;
  cudf::table out1;
  ...
  // Do stuff with out0, out1
  
  // Return a std::pair of the two outputs
  return std::make_pair(std::move(out0), std::move(out1));
}

cudf::table out0;
cudf::table out1;
std::tie(out0, out1) = cudf::return_two_outputs();
```

Note:  `std::tuple`  _could_  be used if not for the fact that Cython does not support 

> `std::tuple`. Therefore, libcudf APIs must use  `std::pair`, and are therefore limited to return 

Reviewer: ```suggestion
`std::tuple`. Therefore, libcudf APIs must use `std::pair`, and are therefore limited to return 
```

# libcudf C++ Developer Guide

This document serves as a guide for contributors to libcudf C++ code. Developers should also refer 
to these additional files for further documentation of `libcudf` best practices.

* [Documentation Guide](DOCUMENTATION.md) for guidelines on documenting libcudf code.

> * [Testing Guide(TESTING.md) for guidelines on writing unit tests.

Reviewer 3: ```suggestion
* [Testing Guide](TESTING.md) for guidelines on writing unit tests.
```

* [Benchmarking Guide](TODO) for guidelines on writing unit benchmarks.

# Overview

libcudf is a C++ library that provides GPU-accelerated data-parallel algorithms for processing 

> column-oriented tabular data. Algorithms provided range from slicing and filtering to sorting, 
> and various types of aggregations, to database type operations such as grouping and joins. libcudf
> serves a number of clients via multiple language interfaces, including Python and Java. Users may

Reviewer 3: The structure of this sentence is a bit odd "From x & y TO z & A TO types of B & C"
```suggestion
column-oriented tabular data. Algorithms provided range from slicing, filtering, sorting 
and various types of aggregations, to database type operations such as grouping and joins. libcudf
serves a number of clients via multiple language interfaces, including Python and Java. Users may
```

structures used in libcudf. Most libcudf algorithms operate on columns. Columns may have a validity
mask representing whether each element is valid or null (invalid). Columns of nested types are 
supported, meaning that a column may have child columns.

### Element

An individual data item within a column. Also known as a row.

### Scalar

A type representing a single element of a data type.

### Table

A table is a collection of columns. It is also known as a dataframe.

### View

A view is a non-owning object that provides zero-copy access (possibly with slicing or offsets) data 
owned by another object. Examples are column views and table views. 

# Directory Structure and File Naming

External/public libcudf APIs are grouped based on functionality into an appropriately titled 
header file  in `cudf/cpp/include/cudf/`. For example, `cudf/cpp/include/cudf/copying.hpp` 
contains the APIs for functions related to copying from one column to another. Note the  `.hpp` 
file extension used to indicate a C++ header file.

Header files should use the `#pragma once` include guard. 

The naming of external API headers should be consistent with the name of the folder that contains 

> the source files that implement the API. For example, the implementation of the APIs found in  
> `cudf/cpp/include/cudf/copying.hpp` are located in `cudf/src/copying`. Likewise, the unit tests for 

Reviewer 3: There is an unnecessary line break here
```suggestion
the source files that implement the API. For example, the implementation of the APIs found in `cudf/cpp/include/cudf/copying.hpp` are located in `cudf/src/copying`. Likewise, the unit tests for 
```

the APIs reside in `cudf/tests/copying/`.

## File extensions

- `.hpp` : C++ header files
- `.cpp` : C++ source files
- `.cu`  : CUDA C++ source files
- `.cuh` : Headers containing CUDA device code

Only use `.cu` and `.cuh` if necessary. A good indicator is the inclusion of `__device__` and other
symbols that are only recognized by `nvcc`. Also, Thrust APIs executed on a device execution policy,

> which should always be `rmm::exec_policy` in libcudf.

Reviewer 3: ```suggestion
which should always use `rmm::exec_policy` in libcudf.
```

# libcudf Data Structures

Application data in libcudf is contained in Columns and Tables, but there are a variety of other
data structures you will use when developing libcudf code.

## Views and Ownership

Resource ownership is an essential concept in `libcudf`. In short, an "owning" object owns a 
resource (such as device memory). It acquires that resource during construction and releases the 

> resource in destruction (RAII). A "non-owning" object does not own resources. Any class in `libcudf` 

Reviewer 3: ```suggestion
resource in destruction ([RAII](https://en.cppreference.com/w/cpp/language/raii)). A "non-owning" object does not own resources. Any class in `libcudf` 
```

and its validity from the device. This can be obtained using the function 
`get_scalar_device_view(ScalarType s)`. Note that a device view is not provided for a base scalar 
object, only for the derived typed scalar class objects.

# libcudf++ API and Implementation

## Streams

CUDA streams are not yet exposed in external libcudf APIs. However, in order to ease the transition 
to future use of streams, all libcudf APIs that allocate device memory or execute a kernel should be 
implemented using asynchronous APIs on the default stream (e.g., stream 0).

The recommended pattern for doing this is to make the definition of the external API invoke an 
internal API in the `detail` namespace. The internal `detail` API has the same parameters as the 
public API, plus a `rmm::cuda_stream_view` parameter at the end defaulted to 
`rmm::cuda_stream_default`. The implementation should be wholly contained in the `detail` API 
definition and use only asynchronous versions of CUDA APIs with the stream parameter.

In order to make the `detail` API callable from other libcudf functions, it should be exposed in a 

> header placed in the `cudf/cpp/include/detail/` directory.

Reviewer 2: Seems we should mention this in the Directory Structure section as well?

Author: It's mentioned in the section on `detail` namespace, but added a sentence about where these detail headers should live to the Directory Structure section.

  ..., rmm::device_memory_resource * mr = rmm::mr::get_current_device_resource());

// This API does not allocate any new *output* memory, therefore
// a memory resource is unnecessary
void does_not_allocate_output_memory(...);                                              
```

### Temporary Memory

Not all memory allocated within a libcudf API is returned to the caller. Often algorithms must 
allocate temporary, scratch memory for intermediate results. Always use the default resource
obtained from `rmm::mr::get_current_device_resource()` for temporary memory allocations. Example:

```c++
rmm::device_buffer some_function(
  ..., rmm::mr::device_memory_resource mr * = rmm::mr::get_current_device_resource()) {
    rmm::device_buffer returned_buffer(..., mr); // Returned buffer uses the passed in MR
    ...
    rmm::device_buffer temporary_buffer(...); // Temporary buffer uses default MR
    ...
    return returned_buffer;
}
```

### Memory Management

libcudf code generally eschews raw pointers and direct memory allocation. Use RMM classes built to
use `device_memory_resource`(*)s for device memory allocation with automated lifetime management.

> #### `rmm::device_buffer`

Reviewer 2: Maybe these classes should be defined in the rmm repo developer guide and then just pointed to from here?

Author: We don't have such a guide yet. But when we do...

![strings](strings.png)

## Structs columns

Structs are represented similarly to lists, except that they have multiple child data columns.
The parent column's type is `STRUCT` and contains no data, but its size represents the number of 
structs in the column, and its null mask represents the validity of each struct element. The parent 
has `N + 1` children, where `N` is the number of fields in the struct. 

1. A non-nullable column of `INT32` elements that indicates the offset to the beginning of each 
   struct in each dense column of elements.
2. For each field, a column containing the actual field data and optional null mask for all elements
   of all the structs packed together.
   
With this representation, `child[0][offsets[i]]` is the first field of struct `i`, 
`child[1][offsets[i]]` is the second field of struct `i`, etc.

## Dictionary columns

Dictionaries provide an efficient way to represent low-cardinality data by storing a single copy 
of each value. A dictionary comprises a column of sorted keys and a column containing an index into the 
keys column for each row of the parent column. The keys column may have any libcudf data type, 

> such as a numerical type or strings. The indices represent the corresponding positions of each 
> element's value in the keys.

Reviewer 2: ```suggestion
such as a numerical type or strings. The indices are represent the corresponding positions of each 
element's value in the keys. The indices child column can be any unsigned integer type (UINT8, UINT16, UINT32, or UINT64).
```

Author: I don't think the added `are` is correct.

Author: Added the sentence about index types.

* [Testing Guide](TESTING.md) for guidelines on writing unit tests.
* [Benchmarking Guide](TODO) for guidelines on writing unit benchmarks.

# Overview

libcudf is a C++ library that provides GPU-accelerated data-parallel algorithms for processing 
column-oriented tabular data. libcudf provides algorithms including slicing, filtering, sorting, 
various types of aggregations, and database-type operations such as grouping and joins. libcudf
serves a number of clients via multiple language interfaces, including Python and Java. Users may
also use libcudf directly from C++ code.

## Lexicon

This section defines terminology used within libcudf

### Column

A column is an array of data of a single type. Along with Tables, columns are the fundamental data 
structures used in libcudf. Most libcudf algorithms operate on columns. Columns may have a validity
mask representing whether each element is valid or null (invalid). Columns of nested types are 
supported, meaning that a column may have child columns.

### Element

An individual data item within a column. Also known as a row.

### Scalar

A type representing a single element of a data type.

### Table

> A table is a collection of columns. It is also known as a dataframe.

Reviewer 4: > the C++ corollary to a DataFrame...

That is very interesting phrasing. I would've have prosaically termed it as "equivalent" to a `DataFrame`.

> A table is a collection of columns. It is also known as a dataframe.

Reviewer 3: i was just copying wording from later in the doc

> A table is a collection of columns. It is also known as a dataframe.

Author: Hmmm, "corollary" doesn't appear later in the doc...

> A table is a collection of columns. It is also known as a dataframe.

Author: I also wouldn't CamelCase or capitalize "data frame".  I decided to go with "equivalent", and fix spelling to "data frame".

### View

> A view is a non-owning object that provides zero-copy access (possibly with slicing or offsets) data 

Reviewer 5: I think a section on slicing and column_view offsets is warranted.  It's an easy-to-overlook case and the word "offset" is heavily overloaded in cudf.

Author: Added

libcudf code uses [snake_case](https://en.wikipedia.org/wiki/Snake_case) for all names except in a 
two cases: template parameters and unit tests and test case names may use Pascal case, aka 
[UpperCamelCase](https://en.wikipedia.org/wiki/Camel_case). We do not use [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation), except sometimes when naming device data variables and their corresponding
host copies. Private member variables are typically prefixed with an underscore.

```c++
template <typename IteratorType>
void algorithm_function(int x, rmm::cuda_stream_view s, rmm::device_memory_resource* mr) 
{
  ...
}

class utility_class 
{
  ...
 private:
  int _rating{};
  std::unique_ptr<cudf::column> _column{};
}

TYPED_TEST_CASE(RepeatTypedTestFixture, cudf::test::FixedWidthTypes);

TYPED_TEST(RepeatTypedTestFixture, RepeatScalarCount)
{
  ...
}
```

C++ formatting is enforced using `clang-format`. You should configure `clang-format` on your 
machine to use the `cudf/cpp/.clang-format` configuration file, and run `clang-format` on all 
changed code before committing it. The easiest way to do this is to configure your editor to 
"format on save".

> 

Reviewer 2: How about some guidelines for `#include` statements? 
From @harrism slack message in November:

1. Group your includes by library (e.g. cuDF, RMM, Thrust, STL)
2. Separate groups by a blank line.
3. Order the groups from "nearest" to "farthest"
4. Use `<>` instead of `""` unless the header is in the same directory as the source file.
5. Let clang-format sort the headers within the groups
6. If you remove code, check whether that makes some includes obsolete, and remove them if so.

Maybe 6 can be generalized with something like:
"Always check the included files are only necessary for the file in which they are included. Try to avoid excessive including especially in header files."

Author: Added.

Aspects of code style not discussed above and not automatically enforceable are typically caught
during code review, or not enforced.

> 

Reviewer 2: Anything else we should include here?

- Usage of `static_cast` and `reinterpret_cast` instead of cast operator `(type)`
- Avoiding the comma operator https://github.com/rapidsai/cudf/pull/7090#discussion_r553330672
- Prefer using thrust or STL algorithms over for-loops when possible.

Author: I could just link the C++ core guidelines and Sean Parent's C++ seasoning talk and say "commit this to memory". :)

Author: Added a C++ Guidelines subsection. Thanks!

public API, plus a `rmm::cuda_stream_view` parameter at the end defaulted to 
`rmm::cuda_stream_default`. The implementation should be wholly contained in the `detail` API 
definition and use only asynchronous versions of CUDA APIs with the stream parameter.

In order to make the `detail` API callable from other libcudf functions, it should be exposed in a 
header placed in the `cudf/cpp/include/detail/` directory.

For example:

```c++
// cpp/include/cudf/header.hpp
void external_function(...);

// cpp/include/cudf/detail/header.hpp
namespace detail{
void external_function(..., rmm::cuda_stream_view stream = rmm::cuda_stream_default)
} // namespace detail

// cudf/src/implementation.cpp
namespace detail{
    // defaulted stream parameter
    void external_function(..., rmm::cuda_stream_view stream){
        // implementation uses stream w/ async APIs
        rmm::device_buffer buff(...,stream);
        CUDA_TRY(cudaMemcpyAsync(...,stream.value()));
        kernel<<<..., stream>>>(...);
        thrust::algorithm(rmm::exec_policy(stream), ...);
    }
} // namespace detail

void external_function(...){
    detail::external_function(...);
}
```

> **Note:** It's important to synchronize the stream if *and only if* it is necessary. For example,

Reviewer 4: ```suggestion
**Note:** It is important to synchronize the stream if *and only if* it is necessary. For example,
```

when a non-pointer value is returned from the API that is the result of an asynchronous 
device-to-host copy, the stream used for the copy should be synchronized before returning. However,
when a column is returned, the stream should not be synchronized because doing so will break 
asynchrony if and when we add an asynchronous API to libcudf.

**Note:** `cudaDeviceSynchronize()` should *never* be used.
 This limits the ability to do any multi-stream/multi-threaded work with libcudf APIs.

 ### Stream Creation

There may be times in implementing libcudf features where it would be advantageous to use streams 
*internally*, i.e., to accomplish overlap in implementing an algorithm. However, dynamically 

> creating a stream can be expensive. RMM has recently added a stream pool for this situation to 

Reviewer 4: ```suggestion
creating a stream can be expensive. RMM has a stream pool for this situation to 
```

Reviewer 4: Hmm... Please feel free to ignore this suggestion. I wanted to avoid references to points in time, to prevent said reference from going out of date. But I see that the following sentence refers to this as `not yet exposed in libcudf`.

Author: I generally agree with your sentiment here. I'll improve it.

avoid dynamic stream creation. However, this is not yet exposed in libcudf, so for the time being, 
libcudf features should avoid creating streams (even if it is slightly less efficient). It is a 

> good idea to leave a `// TODO:` note indicating where using a stream would be beneficial.

Reviewer 4: That `TODO` is intentional and valid, methinks.

non-trivially copyable objects (and required for types with deleted copy constructors, like 
`std::unique_ptr`).

```c++
std::pair<table, table> return_two_tables(void){
  cudf::table out0;
  cudf::table out1;
  ...
  // Do stuff with out0, out1
  
  // Return a std::pair of the two outputs
  return std::make_pair(std::move(out0), std::move(out1));
}

cudf::table out0;
cudf::table out1;
std::tie(out0, out1) = cudf::return_two_outputs();
```

Note:  `std::tuple`  _could_  be used if not for the fact that Cython does not support 
`std::tuple`. Therefore, libcudf APIs must use  `std::pair`, and are therefore limited to return 
only two objects of different types. Multiple objects of the same type may be returned via a 
`std::vector<T>`.

## Iterator-based interfaces

Increasingly, libcudf is moving toward internal (`detail`) APIs with iterator parameters rather 
than explicit `column`/`table`/`scalar` parameters. As with STL, iterators enable generic 
algorithms to be applied to arbitrary containers. A good example of this is `cudf::copy_if_else`. 
This function takes two inputs, and a Boolean mask. It copies the corresponding element from the 

> first or second input depending on whether the mask at that index is true or false. Implementing

Reviewer 3: ```suggestion
first or second input depending on whether the mask at that index is `true` or `false`. Implementing
```

`copy_if_else` for all combinations of `column` and `scalar` parameters is simplified by using
iterators in the `detail` API.

```c++
template <typename FilterFn, typename LeftIter, typename RightIter>
std::unique_ptr<column> copy_if_else(
  bool nullable,
  LeftIter lhs_begin,
  LeftIter lhs_end,
  RightIter rhs,
  FilterFn filter,
  ...);
```
`LeftIter` and `RightIter` need only implement the necessary interface for an iterator. libcudf 
provides a number of iterator types and utilities that are useful with iterator-based APIs from 
libcudf as well as Thrust algorithms. Most are defined in `include/detail/iterator.cuh`. 

### Pair iterator

The pair iterator is used to access elements of nullable columns as a pair containing an element's 
value and validity. `cudf::detail::make_pair_iterator` can be used to create a pair iterator from a 

> `column_device_view` or a `cudf::scalar`.

Reviewer 4: Should we mention explicitly that `make_pair_iterator` is not available on `mutable_column_device_view`?

### Null-replacement iterator

> This iterator replaces the null/validity value for each element with a specified constant (true or
> false). Created using `cudf::detail::make_null_replacement_iterator`.
> 
> ### Validity iterator
> 
> This iterator returns the validity of the underlying element (true or false). Created using 

Reviewer 3: ```suggestion
This iterator replaces the null/validity value for each element with a specified constant (`true` or
`false`). Created using `cudf::detail::make_null_replacement_iterator`.

### Validity iterator

This iterator returns the validity of the underlying element (`true` or `false`). Created using 
```

The "Indexalator", or index-normalizing iterator (`include/cudf/detail/indexalator.cuh`), can be 
used for index types (integers) without requiring a type-specific instance. It can be used for any 
iterator interface for reading an array of integer values of type `int8`, `int16`, `int32`, 
`int64`, `uint8`, `uint16`, `uint32`, or `uint64`. Reading specific elements always return a 
`cudf::size_type` integer.

Use the `indexalator_factory` to create an appropriate input iterator from a column_view. Example 
input iterator usage:

```c++
auto begin = indexalator_factory::create_input_iterator(gather_map);
auto end   = begin + gather_map.size();
auto result = detail::gather( source, begin, end, IGNORE, stream, mr );
```

Example output iterator usage:

```c++
auto result_itr = indexalator_factory::create_output_iterator(indices->mutable_view());
thrust::lower_bound(rmm::exec_policy(stream),
                    input->begin<Element>(),
                    input->end<Element>(),
                    values->begin<Element>(),
                    values->end<Element>(),
                    result_itr,
                    thrust::less<Element>());
```

## Namespaces

### External

> All public libcudf APIs should be placed in the `cudf` namespace*. Example:

Reviewer 3: ```suggestion
All public libcudf APIs should be placed in the `cudf` namespace. Example:
```

    switch(t.id())
       case type_id::INT32: f.template operator()<int32_t>()
       case type_id::INT64: f.template operator()<int64_t>()
       case type_id::FLOAT: f.template operator()<float>()
       ...
}
```

The following example shows a function object called `size_of_functor` that returns the size of the 
dispatched type.

```c++
struct size_of_functor{
  template <typename T>
  int operator()(){ return sizeof(T); }
};

cudf::type_dispatcher(data_type{type_id::INT8}, size_of_functor{});  // returns 1
cudf::type_dispatcher(data_type{type_id::INT32}, size_of_functor{});  // returns 4
cudf::type_dispatcher(data_type{type_id::FLOAT64}, size_of_functor{});  // returns 8
```

By default, `type_dispatcher` uses `cudf::type_to_id<t>` to provide the mapping of `cudf::type_id` 
to dispatched C++ types. However, this mapping may be customized by explicitly specifying a 
user-defined trait for the `IdTypeMap`. For example, to always dispatch `int32_t` for all values of 
`cudf::type_id`:

```c++
template<cudf::type_id t> struct always_int{ using type = int32_t; }

// This will always invoke `operator()<int32_t>`
cudf::type_dispatcher<always_int>(data_type, f);
```

> 

Reviewer 2: I feel we need a Spider-Man quote here: _With great power comes great responsibility._

Try to avoid multiple type-dispatch if possible. The compiler will create a code path for every type so a 2nd type-dispatch would multiply the paths by the number of types. This can create very large object code and cause very long compile times.

Author: Added

The following example functor prints `integral` or `floating point` for integral or floating point
types, respectively.

```c++
struct integral_or_floating_point {
template <typename ColumnType,
          std::enable_if_t<not std::is_integral<ColumnType>::value and
                           not std::is_floating_point<ColumnType>::value>* = nullptr> 
void operator()() { std::cout << "neither integral nor floating point\n"; }

template <typename ColumnType,
          std::enable_if_t<std::is_integral<ColumnType>::value>* = nullptr>
void operator()() { std::cout << "integral\n"; }

template < typename ColumnType,
           std::enable_if_t<std::is_floating_point<ColumnType>::value>* = nullptr> 
void operator()() { std::cout << "floating point\n"; }
};
```

For more info on SFINAE with `std::enable_if`, [see this post](https://eli.thegreenplace.net/2014/sfinae-and-enable_if).

There are a number of traits defined in `include/cudf/utilities/traits.hpp` that are useful for 
partial specialization of dispatched function objects. For example `is_numeric<T>()` can be used to 
specialize for any numeric type.

# Nested Data Types

> libcudf supports a number of nested data types, including strings, lists, and structs. A string is 

Reviewer 4: I wonder if we are conflating "variable size" types with "nested types" here:
1. Are strings really nested types?
2. `struct` is only of variable size if it (eventually) contains a string or list.

Reviewer 5: >     1. Are strings really nested types?

This is a really slippery question.  To me, they are in limbo between nested and non-nested.  

The string column itself does indeed have child columns "nested" inside but that's more of an implementation detail than a data type. It's still logically just "strings".   On the other hand, 1 level list columns are exactly the same and those are clearly nested types.  

On the other, _other_ hand, I've found that when writing tests, I tend to only use the term "Nested" in the name when the test in question has >= two logical columns nested inside each other (ie, List<List<int>>).  If I'm just making a test with a basic list or struct, I tend to use "List" or "Struct" in the name.  

Arrow probably has something to say about this.

Author: I rephrased to "Variable-size and Nested Data Types".

simply a character string, but a column of strings may have a different-length string in each row.
A list is a list of elements of any type, so a column of lists of integers has rows with a list of 

> integers, possibly of a different length, in each row. In a colum of structs, each row is a 

Reviewer 4: ```suggestion
integers, possibly of a different length, in each row. In a column of structs, each row is a 
```

> libcudf supports a number of nested data types, including strings, lists, and structs. A string is 
> simply a character string, but a column of strings may have a different-length string in each row.
> A list is a list of elements of any type, so a column of lists of integers has rows with a list of 
> integers, possibly of a different length, in each row. In a colum of structs, each row is a 
> structure comprising one or more fields. These fields are stored in structure-of-arrays format, so 
> that the column of structs has a nested column for each field of the structure. 

Reviewer 4: Might we consider converting this paragraph to a list? (No pun intended.) 
Also, fixed typo in `colum`.
```suggestion
libcudf supports a number of nested data types, including strings, lists, and structs: 
1. `String`: This is simply a character string, but a column of strings may have a 
different-length string in each row.
2. `List`: This is a list of elements of any type, so a column of lists of integers has 
rows with a list of integers, possibly of a different length, in each row. 
3. `Struct`: In a column of structs, each row is a structure comprising one or more 
fields. These fields are stored in structure-of-arrays format, so that the column of 
structs has a nested column for each field of the structure. 
```

experience. Therefore it is important to carefully write algorithms, and to test and document them
well.

## List columns

In order to represent variable-width elements, libcudf columns contain a vector of child columns.
For list columns, the parent column's type is `LIST` and contains no data, but its size represents
the number of lists in the column, and its null mask represents the validity of each list element.
The parent has two children. 

1. A non-nullable column of `INT32` elements that indicates the offset to the beginning of each list
   in a dense column of elements.
2. A column containing the actual data and optional null mask for all elements of all the lists 
   packed together.
   
With this representation, `data[offsets[i]]` is the first element of list `i`, and the size of list
`i` is given by `offsets[i+1] - offsets[i]`.

Note that the data may be of any type, and therefore the data column may itself be a nested column
of any type. Note also that not only is each list nullable (using the null mask of the parent), but
each list element may be nullable. So you may have a lists column with null row 3, and also null
element 2 of row 4.

> 

Reviewer 5: A key point with list columns is:
- The actual underlying data for a lists column is always bundled together into a single leaf column at the very bottom of the hierarchy (ignoring structs, which conceptually "reset" the root of the hierarchy), regardless of the level of nesting.  So a `List<List<List<List<int>>>>>` column still just has a single int column at the very bottom.  

A semi-visual representation might be useful here:

```
   lists_column = { {{{1, 2}, {3, 4}}, NULL}, {{{10, 20}, {30, 40}}, {{50, 60, 70}, {0}}} }

   List<List<List<int>>>  (2 rows):
   Length : 2
   Offsets : 0, 2, 4
   Children :
      List<List<int>>:
      Length : 4
      Offsets : 0, 2, 2, 4, 6
      Null count: 1
        1101
      Children :
        List<int>:
        Length : 6
        Offsets : 0, 2, 4, 6, 8, 11, 12
        Children :
          Column of ints
          1, 2, 3, 4, 10, 20, 30, 40, 50, 60, 70, 0
```

Reviewer 5: A link to the arrow docs would be good too:

https://arrow.apache.org/docs/format/Columnar.html?highlight=nested%20types#physical-memory-layout

Our list implementation is specifically the "Variable-size List" physical layout.

Author: Added all of this. Thanks.

## Strings columns

Strings are represented in much the same way as lists, except that the data child column is always 
a non-nullable column of `INT8` data. The parent column's type is `STRING` and contains no data,
but its size represents the number of strings in the column, and its null mask represents the 
validity of each string. To summarize, the strings column children are:

1. A non-nullable column of `INT32` elements that indicates the offset to the beginning of each 
   string in a dense column of all characters.
2. A non-nullable column of `INT8` elements of all the characters across all the strings packed 
   together.

With this representation, `characters[offsets[i]]` is the first character of string `i`, and the 
size of string `i` is given by `offsets[i+1] - offsets[i]`. The following image shows an example of
this compound column representation of strings.

![strings](strings.png)

## Structs columns

Structs are represented similarly to lists, except that they have multiple child data columns.
The parent column's type is `STRUCT` and contains no data, but its size represents the number of 

> structs in the column, and its null mask represents the validity of each struct element. The parent 

Reviewer 4: I'm trying to decide whether the interplay between the validity masks of structs and its members should be described here. Specifically that the parent's validity mask is bitwise-ANDed with all its children's masks, so that a null `struct` row at the parent level marks the corresponding child row to null, for all child columns.

Would an explanation (with examples)  make this section too verbose?

Author: It might, but it's better to be too verbose and solve that problem by refactoring than to have missing information. If you can provide something in a suggestion comment, I'll accept it.

Author: I'm not sure I would get this written correctly by myself. Would be great if you could write it, either in this PR or file an issue after and do it as a followup.

Reviewer 4: Sorry I missed this. I'll add a section shortly.

Reviewer 4: Hmm... I can't seem to add a `suggestion` to this diff. 
Perhaps something like this? (Will likely need rephrasing.)
Please note that the "\\```" will need the "\\" removed. I couldn't figure out how to escape this correctly in the suggestion text. :/

```
As per the [Apache Arrow specification](https://arrow.apache.org/docs/format/Columnar.html#struct-layout), in addition to the struct column's null mask, each struct
field column has its own optional null mask. A struct field's validity can vary independently from the 
corresponding struct row. For instance, a non-null struct row might have a null field. However, the fields
of a null struct row are deemed to be null as well.
As an example, consider a struct column of type `STRUCT<FLOAT32, INT32>`. If the contents were
`[ {1.0, 2}, {4.0, 5}, null, {8.0, null} ]`, the struct column's layout would be as follows:

\```
{
  type = STRUCT
  null_mask = 1011
  null_count = 1
  children = {
    {   
      type = FLOAT32
      data = [1.0, 4.0, X, 8.0]
      null_mask  = 1011
      null_count = 1
    },  
    {   
      type = INT32
      data = [2, 5, X, X]
      null_mask  = 0011
      null_count = 2
    }  
  }   
}
\```
The last struct row (index 3) is not null, but has a null value for its INT32 field. This is 
reflected in bit 3 of the null masks of the struct column and the INT32 field.
Also, row 2 of the struct column is null. Its corresponding fields should also be null, so bit 2 is
unset in the null masks of both struct fields.

```

Reviewer 2: I believe the bit indices are shown right-to-left `76543210`. 
So `1011` means that only `row0`, `row1`, and `row3` are valid.

Reviewer 4: 1 == valid. 0 == null.
I based the null-mask representation on how it shows up when the column is printed. I realize that this is confusing.

I have modified the example to show the null_mask as an array instead. This should be clearer.

Reviewer 4: @davidwendt, et al, 

Would anyone know how one makes a "```suggestion" block that itself contains a "```" code block, in GitHub? I'd like to make my suggestion above available to pick into this PR, but I'm thwarted by GitHub's formatting. :/

Reviewer 2: It looks like you may be able to change the 3 backtick marks to 3 tildes for the suggestion block.
`````
~~~suggestion
Some text.
```
Example text.
```
Other text.
~~~
`````

I think it is also possible to nest blocks by using more than 3 backticks.

Author: @mythrocks the problem I have with your example is this: 

> The last struct row (index 3) is not null, but has a null value in the INT32 field. This is reflected in bit 3 of the null masks of the struct column and the INT32 field.

But the column-level (parent-level?) null mask is [1, 1, 0, 1] (bit 3 is valid), while the INT32 field null mask is [1, 1, 0, 0] (bit 3 is invalid/null).  So your text does not match your example.

Reviewer 4: >> The last struct row (index 3) is not null...

> But the column-level (parent-level?) null mask is [1, 1, 0, 1] (bit 3 is valid)

A) `1` indicates valid. In the `STRUCT` mask, bit 3 is 1. Therefore, `STRUCT` at index 3 is valid, i.e. not null. The two quotes above are not at odds.

>> but has a null value in the INT32 field...

> while the INT32 field null mask is [1, 1, 0, 0] (bit 3 is invalid/null).

B) So the `INT32` field's mask[3] is 0, i.e. invalid.

From A and B above, the `STRUCT` is valid, but its `INT32` field is invalid. Hence, `{8.0, ∅}`.

Forgive me, I'm still unable to see the problem. :/

Author: Sorry, I read 

> This is reflected in bit 3 of the null masks of the struct column and the INT32 field.

To mean that the null masks of the struct column *and* the INT32 field should be null at bit 3. I will rewrite this sentence to be clear, it's unclear currently.

Reviewer 4: Apologies for the clumsy wording. I meant that "this" is reflected in bit 3 of the null masks of the `STRUCT` column and the `INT32` field (i.e. `1` and `0` respectively).

For this reason, it is recommended to use the `string_view` class to access these characters for
most operations.

The `string_view.cuh` header also includes some utility methods for reading and writing 
(`to_char_utf8/from_char_utf8`) individual UTF-8 characters to/from byte arrays.

### `cudf::lists_column_view` and `cudf::lists_view`

`cudf::lists_column_view` is a view of a lists column. `cudf::list_view` is a view of a single list, and 
therefore `cudf::list_view` is the data type of a `cudf::column` of type `LIST`.

`cudf::type_dispatcher` dispatches to the `list_view` data type when invoked on a `LIST` column.

### `cudf::structs_column_view` and `cudf::struct_view`

`cudf::structs_column_view` is a view of a structs column. `cudf::struct_view` is a view of a single
struct, and therefore `cudf::struct_view` is the data type of a `cudf::column` of type `STRUCT`.

`cudf::type_dispatcher` dispatches to the `struct_view` data type when invoked on a `STRUCT` column.

# cuIO: file reading and writing

cuIO is a component of libcudf that provides GPU-accelerated reading and writing of data file 

> formats commonly used in data analytics, including CSV, Parquet, ORC, and Avro.

Reviewer: and JSON Lines (not sure if omitted intentionally).

Author: Added

header file  in `cudf/cpp/include/cudf/`. For example,  `cudf/cpp/include/cudf/copying.hpp` 
contains the APIs for functions related to copying from one column to another. Note the  `.hpp` 
file extension used to indicate a C++ header file. 

Header files should use the `#pragma once` include guard. 

The naming of external API headers should be consistent with the name of the folder that contains 
the source files that implement the API. For example, the implementation of the APIs found in  
`cudf/cpp/include/cudf/copying.hpp` are located in `cudf/src/copying`. Likewise, the unit tests for 
the APIs reside in `cudf/tests/copying/`.

# libcudf Data Structures

## Views and Ownership

Resource ownership is an essential concept in `libcudf`. In short, an "owning" object owns a 
resource (such as device memory). It acquires that resource during construction and releases the 
resource in destruction (RAII). A "non-owning" object does not own resources. Any class in `libcudf` 
with the `*_view` suffix is non-owning. For more detail see the 
[`libcudf++` presentation.](https://docs.google.com/presentation/d/1zKzAtc1AWFKfMhiUlV5yRZxSiPLwsObxMlWRWz_f5hA/edit?usp=sharing)

> `libcudf` functions typically take views as input (`column_view`, `table_view`, or `scalar_view`)

Reviewer 6: ```suggestion
`libcudf` functions typically take views as input (`column_view` or `table_view`)
```

Author: We don't have `scalar_view`?

and produce `unique_ptr`s to owning objects as output. For example, 

```c++

> std::unique_ptr<table> sort(table_view input);

Reviewer 6: ```suggestion
std::unique_ptr<table> sort(table_view const& input);
```

```

## `rmm::device_memory_resource`<a name="memory_resource"></a>

> `libcudf` Allocates all memory via RMM memory resources (MR). See the 

Reviewer 6: ```suggestion
`libcudf` Allocates all device memory via RMM memory resources (MR). See the 
```

The recommended pattern for doing this is to make the definition of the external API invoke an internal API in the `detail` namespace. The internal `detail` API will have all the same parameters, plus a `rmm::cuda_stream_view` parameter at the end defaulted to `rmm::cuda_stream_default`. 
The implementation should be wholly contained in the `detail` API definition and use only asynchronous versions of CUDA APIs with the defaulted stream parameter. 

In order to make the `detail` API callable from other libcudf functions, it should be exposed in a header placed in the `cudf/cpp/include/detail/` directory.

For example:

```c++
// cpp/include/cudf/header.hpp
void external_function(...);

// cpp/include/cudf/detail/header.hpp
namespace detail{
void external_function(..., rmm::cuda_stream_view stream = rmm::cuda_stream_default)
} // namespace detail

// cudf/src/implementation.cpp
namespace detail{
    // defaulted stream parameter
    void external_function(..., rmm::cuda_stream_view stream){
        // implementation uses stream w/ async APIs

>         RMM_ALLOC(...,stream);

Reviewer 6: yikes
```suggestion
        rmm::device_buffer buff(...,stream);
```

Author: Note that I have only edited the very beginning of this document so far. This is a WIP PR not really ready for review.

It *is* ready for you to push directly to the PR though if you would like.

        CUDA_TRY(cudaMemcpyAsync(...,stream.value()));
        kernel<<<..., stream>>>(...);
        thrust::algorithm(rmm::exec_policy(stream), ...);

>         stream.synchronize();

Reviewer 6: This sync may lead people to believe that all detail functions should sync their stream before returning.

>         RMM_FREE(...,stream);

Reviewer 6: ```suggestion
```

// therefore the API must accept a memory resource pointer
std::unique_ptr<column> returns_output_memory(..., 
                                              rmm::device_memory_resource * mr = rmm::mr::get_current_device_resource());

// This API does not allocate any new *output* memory, therefore
// a memory resource is unnecessary
void does_not_allocate_output_memory(...);                                              
```

### Temporary Memory

Not all memory allocated within a libcudf feature is returned to the caller.
Oftentimes in implementing an algorithm, it is necessary to allocate temporary, scratch memory for intermediate results.
For these temporary memory allocations, the default resource returned from `rmm::mr::get_current_device_resource()` should *always* be used. Example:
```c++
rmm::device_buffer some_function(..., rmm::mr::device_memory_resource mr * = rmm::mr::get_current_device_resource()){
    rmm::device_buffer returned_buffer(..., mr); // Returned buffer uses the passed in MR
    ...
    rmm::device_buffer temporary_buffer(...); // Temporary buffer uses default MR
    ...
    return returned_buffer;
}
```

### Memory Management

> Explicit memory management through calls to `RMM_ALLOC/RMM_FREE` should be avoided whenever possible in favor of a construct with automated lifetime management, i.e., a RAII object. 

Reviewer 6: ```suggestion
```

void * raw_data = buff.data(); // Raw pointer to underlying device memory

rmm::device_buffer copy(buff); // Deep copies `buff` into `copy`
rmm::device_buffer moved_to(std::move(buff)); // Moves contents of `buff` into `moved_to`
```

#### `rmm::device_scalar<T>`
Allocates a single element of the specified type initialized to the specified value.
Use this for scalar input/outputs into device kernels, e.g., reduction results, null count, etc.
This is effectively a convenience wrapper around a `rmm::device_vector<T>` of length 1.

```c++
// Allocates device memory for a single int using the specified resource and stream
// and initializes the value to 42
rmm::device_scalar<int> int_scalar{42, stream, mr}; 

// scalar.data() returns pointer to value in device memory
kernel<<<...>>>(int_scalar.data(),...);

// scalar.value() synchronizes the scalar's stream and copies the 
// value from device to host and returns the value
int host_value = int_scalar.value();
```

#### `rmm::device_vector<T>`

Allocates a specified number of elements of the specified type.
If no initialization value is provided, all elements are default initialized (this incurs a kernel launch).

> (*) Note: `rmm::device_vector<T>` is not yet updated to use `device_memory_resource`s, but support is forthcoming.

Reviewer 6: Should add a section for `device_uvector`.

  cudf::table out0;
  cudf::table out1;
  ...
  // Do stuff with out0, out1
  
  // Return a std::pair of the two outputs
  return std::make_pair(std::move(out0), std::move(out1));
}

cudf::table out0;
cudf::table out1;
std::tie(out0, out1) = cudf::return_two_outputs();
```

Note:  `std::tuple`  _could_  be used if not for the fact that Cython does not support  `std::tuple`. Therefore, libcudf APIs must use  `std::pair`, and are therefore limited to return only two objects of different types. Multiple objects of the same type may be returned via a  `std::vector<T>`.

## Namespaces

### External
All public libcudf APIs should be placed in the `cudf` namespace*. Example:
```c++
namespace cudf{
   void public_function(...);
} // namespace cudf
```

For most functions, the top-level `cudf` namespace is sufficient. However, for logically grouping a broad set of functions, further namespaces may be used. For example, there are numerous functions that are specific to columns of Strings. These functions are put in the `cudf::strings::` namespace. Similarly, functionality used exclusively for unit testing is placed in the `cudf::test::` namespace. 

> ### `experimental`
> During the transition period, symbols in `libcudf++` that conflict with old symbol names should be placed in the `cudf::experimental` namespace to prevent collision with the old symbols, e.g., `cudf::experimental::table` and `cudf::experimental::type_dispatcher`. Once the transition is complete, the `experimental` namespace will be removed.

Reviewer 6: ```suggestion
```

    switch(t.id())
       case type_id::INT32: f.template operator()<int32_t>()
       case type_id::INT64: f.template operator()<int64_t>()
       case type_id::FLOAT: f.template operator()<float>()
       ...
}
```

The below example shows a function object called `size_of_functor` that returns the size of the dispatched type:

```c++
struct size_of_functor{
template <typename T>
int operator()(){ return sizeof(T); }
};
cudf::type_dispatcher(data_type{type_id::INT8}, size_of_functor{});  // returns 1
cudf::type_dispatcher(data_type{type_id::INT32}, size_of_functor{});  // returns 4
cudf::type_dispatcher(data_type{type_id::FLOAT64}, size_of_functor{});  // returns 8
```

By default, the `type_dispatcher` uses `cudf::type_to_id<t>` to provide the mapping
of `cudf::type_id`s to dispatched C++ types.
However, this mapping may be customized by explicitly specifying a user-defined trait for the `IdTypeMap`.
For example, to always dispatch `int32_t` for all values of `cudf::type_id`:

```c++
template<cudf::type_id t> struct always_int{ using type = int32_t; }

// This will always invoke `operator()<int32_t>`
cudf::type_dispatcher<always_int>(data_type, f);
```

> 

Reviewer 6: ```suggestion
## Specializing Type Dispatched Code Paths
```

Example:
```c++
// Creates a non-nullable STRING column with 7 string elements: 
// {"", "this", "is", "a", "column", "of", "strings"}
std::vector<std::string> strings{"", "this", "is", "a", "column", "of", "strings"};
strings_column_wrapper s(strings.begin(), strings.end());

// Creates a nullable STRING column with 7 string elements: 
// {NULL, "this", NULL, "a", NULL, "of", NULL}
std::vector<std::string> strings{"", "this", "is", "a", "column", "of", "strings"};
auto validity = make_counting_transform_iterator(0, [](auto i){return i%2;});
strings_column_wrapper s(strings.begin(), strings.end(), validity);

// Creates a non-nullable STRING column with 7 string elements: 
// {"", "this", "is", "a", "column", "of", "strings"}
strings_column_wrapper s({"", "this", "is", "a", "column", "of", "strings"});

// Creates a nullable STRING column with 7 string elements: 
// {NULL, "this", NULL, "a", NULL, "of", NULL}
auto validity = make_counting_transform_iterator(0, [](auto i){return i%2;});
strings_column_wrapper s({"", "this", "is", "a", "column", "of", "strings"}, validity);
```

> 

Reviewer 6: Should add a section on list_column_wrapper. @nvdbaranec

Author: Done.

> 

Reviewer 6: Should add a section on `fixed_point_column_wrapper`. @codereport

Author: Done.

|       `gdf_is_valid()`       |        `*column_device_view::is_valid()`        |         Verifies existence of bitmask before  attempting to read it.         |
|    `bit_mask::is_valid()`    |    `*column_device_view::is_valid_nocheck()`    | Does not verify the existence of the bitmask  before attempting to read it.  |
| `gdf_vald_allocation_size()` |        `bitmask_allocation_size_bytes()`        |                                                                              |
|        `is_nullable()`       |               `*view::nullable()`               |                                                                              |
|         `has_nulls()`        |               `*view::has_nulls()`              |                                                                              |
|        `cudf::copy()`        | `column::column(const&)` `table::table(const&)` |                               Copy constructors                              |
|    `cudf::allocate_like()`   |      `cudf::allocate_like()`      |                      This applies to both columns/tables                     |

> |     `cudf::empty_like()`     |        `cudf::empty_like()`       |                      This applies to both columns/tables                     |

Reviewer 6: All this can be deleted.

## Strings Support<a name="string_support"></a>

> One of the more significant changes in `libcudf++` is the addition of native support for columns of strings. 
> Previously, string support in libcudf was accomplished via the `NVCategory` class.
> See the [NVCategory](#nvcategory_changes) section for more detail. 
> This class will no longer exist in `libcudf++`, but a replacement "dictionary" column type is planned for the future to replace the functionality of NVCategory.
> 
> All libcudf functions will now need to be able to operate on a column of string elements directly. 
> This poses a set of unique challenges, as unlike all previous libcudf types, the elements in a string column are variable width. 

Reviewer 6: ```suggestion

```

> In order to represent variable width strings, libcudf uses a *compound* column (i.e., column with children).

Reviewer 6: There should probably be a previous section generically describing column children. That can then lead into this section on strings. Then add a section on lists and structs.

The `string_view` contains comparison operators `<,>,==,<=,>=` so they can be used in many cudf functions like `sort` without specific strings code.
The data for a `string_view` instance is required to be [UTF-8](#UTF-8) and all operators and methods expect this encoding. Unless documented otherwise,
position and length parameters are specified in characters and not bytes.
The class also includes a `string_view::const_iterator` which can be used to navigate through individual characters within the string.

The `type-dispatcher` dispatches to the `string_view` data type when invoked on a `STRING` column.

### UTF-8

The cudf strings column only supports UTF-8 encoding for strings data. [UTF-8](https://en.wikipedia.org/wiki/UTF-8) is a variable-length character encoding where each character can be 1-4 bytes. This means the length of a string is not the same as its size in bytes. For this reason, it recommended to use the `string_view` class to access these characters for most operations.

The `string_view.cuh` also includes some utility methods for reading and writing (`to_char_utf8/from_char_utf8`) individual UTF-8 characters to/from byte arrays.

> ### NVCategory<a name="nvcategory_changes"></a>
> 
> Previously, libcudf's support for strings columns relied on the `NVCategory` class. 
> `NVCategory` was an object that emulated a *categorical* or *dictionary* column. A dictionary column is a compound column composed of a *dictionary* and *indices* into that dictionary. 
> The dictionary contains the unique set of values within the column.
> The indices contain the offset into the dictionary for each element in the column, e.g., element `i` is given by `dictionary[indices[i]]`.
> See the example below.
> 
> Furthermore, the dictionary is sorted.
> This means the indices are order preserving, and relational comparisons between values in the indices column will provide the same ordering as if the strings were compared directly.
> For example, in the snippet below, `"green"` is ordered lexicographically before `"red"`, and `"green"`'s index is `0` where `"red"`'s is `1`.
> Since `0` is ordered before `1`, comparing `0` and `1` will provide the same ordering as comparing `"green"` and `"red"` directly.
> 
> ```c++
> strings:    {"red", "yellow", "red", "green", "yellow"}
> dictionary: {"green", "red", "yellow"}
> indices:    {1, 2, 1, 0, 2}
> ```
> `libcudf` leveraged the order preserving nature of the indices to allow comparing string elements without ever having to compare strings directly. 
> In effect, this allowed us to treat columns of strings as fixed-width, numeric columns which greatly simplified handling of string columns. 
> 
> However, there was extra work required to manage the `NVCategory` columns--all of which should be removed when porting to libcudf++.
> Any use of a function from `cudf/cpp/include/cudf/utilities/legacy/nvcategory_util.hpp` should be removed from the new implementation.

Reviewer 6: ```suggestion
```
